Add FY4A AGRI L2 reader

zxdawn opened the pull request:
This PR adds a reader for the FY-4A AGRI L2 data which are NC files.

 - [ ] Tests added <!-- for all bug fixes or enhancements -->
 - [ ] Tests passed <!-- for all non-documentation changes -->
 - [ ] Passes ``flake8 satpy`` <!-- remove if you did not edit any Python files -->
 - [ ] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->

zxdawn commented:
Currently, there're several bugs in the reader:

## 1. Wrong plotting results (don't know why)

This is the result of AGRI L1 data () plotted by:

```
import sys
from glob import glob
from satpy import Scene
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
from satpy.utils import debug_on

filenames = glob('FY4A-_AGRI--_N_DISK_1047E_L1-_FDI-_MULT_NOM_20190807060000*')
scn = Scene(reader="agri_l1", filenames=filenames)
scn.load(['C12'])
img = scn['C12']

crs = img.attrs['area'].to_cartopy_crs()
f, ax = plt.subplots(subplot_kw=dict(projection=crs))
ax.gridlines()

plt.imshow(img, transform=crs, extent=crs.bounds, origin='upper', cmap='viridis')
plt.show()
```

![AGRI_L1](https://user-images.githubusercontent.com/30388627/86527709-506ebc00-bed4-11ea-8c95-b360f8c506b0.png)

For L2 CTH data, I used the similar plotting method:
```
filenames = glob('*CTH*NC')
scn = Scene(reader="agri_l2", filenames=filenames)
scn.load(['CTH'])
img = scn['CTH']

crs = img.attrs['area'].to_cartopy_crs()
f, ax = plt.subplots(subplot_kw=dict(projection=crs))
ax.gridlines()

plt.imshow(img, transform=crs, extent=crs.bounds, origin='upper', cmap='viridis')
plt.show()
```

![CTH](https://user-images.githubusercontent.com/30388627/86527727-75fbc580-bed4-11ea-9ed0-423742b05a36.png)

If I plot the `img` directly by `img.plot()`, it looks fine:
![img_plot](https://user-images.githubusercontent.com/30388627/86527750-93309400-bed4-11ea-8c97-dc8927f1f45e.png)

PS: this is the figure generated by Panoply:
![CTH_panoply](https://user-images.githubusercontent.com/30388627/86527763-a3487380-bed4-11ea-93e3-ff7b6d4a8c38.png)


## 2. Only supports data of two resolutions: 1 km and 4 km.

For the other two resolutions (12 km and 64 km), I can't find the proj info like `_COFF_list` from the FY website.
So, I let it raise the error when users use two specific data of these resolutions (AMV: Atmosphere Montion Vector and LSE: Land surface Emissivity).

djhoese commented:
For the plots, what does the colorbar look like on the satpy generated one (the 2nd image)? Do you have any enhancement configured for this product? Wait, what's the difference between how you are generating the 2nd image and the 3rd image?

zxdawn commented:
@djhoese I updated the codes and figures above.

I added the colorbar to images now and didn't apply any enhancement. The second one used `imshow` while the third one used `.plot()`.

djhoese commented:
What if you provide vmin/vmax to the imshow in the second plot?

zxdawn commented:
The second plot used `vmin=0, vmax=20e3`. Sorry for not updating that part in the comment.

zxdawn commented:
I suppose this is caused by the transposed `x` and `y` in L2 product!

Here's the array in the L1 product:
```
dask.array<where, shape=(1108, 2748), dtype=float32, chunksize=(1108, 2748), chunktype=numpy.ndarray>
Dimensions without coordinates: y, x
Attributes:
    band_names:          band2(band number is range from 1 to 14)
    center_wavelength:   0.65um
    long_name:           0.65um channel 4km image data layer
    units:               %
    valid_range:         [ -1.25  101.125]
    platform_name:       FY-4A
    sensor:              agri
    orbital_parameters:  {'satellite_nominal_latitude': 0.0, 'satellite_nomin...
    name:                C02
    wavelength:          0.65 µm (0.55-0.75 µm)
    resolution:          4000
    calibration:         calibration.reflectance
    file_key:            NOMChannel02
    lut_key:             CALChannel02
    fill_value:          65535
    file_type:           agri_l1_4000m
    standard_name:       toa_bidirectional_reflectance
    modifiers:           ()
```

And, the L2 product has the wrong `x` and `y` when compared with it:
```
dask.array<open_dataset-b8e865c599e758ab3330b94e3762c0e6CTH, shape=(1092, 2748), dtype=float32, chunksize=(1092, 2748), chunktype=numpy.ndarray>
Coordinates:
  * y        (y) float32 0.0 1.0 2.0 3.0 4.0 ... 2744.0 2745.0 2746.0 2747.0
  * x        (x) float32 0.0 1.0 2.0 3.0 4.0 ... 1088.0 1089.0 1090.0 1091.0
Attributes:
    long_name:            FY4A PGS AGRI L2 Cloud Top Height
    standard_name:        Cloud Top Height
    _Unsigned:            TRUE
    FillValue:            -999.0
    valid_range:          [1.e+00 2.e+04]
    scale_factor:         1.0
    add_offset:           0.0
    units:                m
    resolution:           4KM
    ancillary_variables:  DQF
    Description:          65535:Space, -999:FillValue
```

The root should be the dimension name stored in the original file.
For L1, they just have the dimension without a name:
```
ushort NOMChannel01(1108, 2748);
  :FillValue = 65535US; // ushort
  :Intercept = 0.0f; // float
  :Slope = 1.0f; // float
  :units = "DN";
  :center_wavelength = "0.47um";
  :band_names = "band1(band number is range from 1 to 14)";
  :long_name = "0.47um channel 4km image data layer";
  :valid_range = 0US, 4095US; // ushort
  :_ChunkSizes = 1108U, 2748U; // uint
```

But, for L2, they assign the name manually:
```
float CTT(x=1108, y=2748);
  :long_name = "FY4A PGS AGRI L2 Cloud Top Temperature";
  :standard_name = "Cloud Top Temperature";
  :_Unsigned = "TRUE";
  :FillValue = -999.0f; // float
  :valid_range = 160.0f, 320.0f; // float
  :scale_factor = 1.0f; // float
  :add_offset = 0.0f; // float
  :units = "K";
  :resolution = "4KM";
  :coordinates = "x y";
  :ancillary_variables = "DQF";
  :Description = "65535:Space, -999:FillValue";
  :_ChunkSizes = 554U, 1374U; // uint

float x(x=1108);
  :long_name = "FY4A fixed grid projection x-coordinate";
  :_ChunkSizes = 1108U; // uint

float y(y=2748);
  :long_name = "FY4A fixed grid projection y-coordinate";
  :_ChunkSizes = 2748U; // uint
```

So, this issue is caused by the wrong stored `x` and `y` in L2 product, right?

zxdawn commented:
Forgot to say, the transposed `x` and `y` could cause error when REGC L2 file is loaded, while DISK file is fine because `x` and `y` are same:
```
................site-packages/xarray/core/dataset.py", line 205, in calculate_dimensions
    % (dim, size, k, dims[dim], last_used[dim])
ValueError: conflicting sizes for dimension 'x': length 1092 on <this-array> and length 2748 on 'x'
```

djhoese commented:
Interesting order of dimensions. I'm not sure I understand why someone would do that. If you force a transpose in the reader in these cases does it change/fix the plots?

zxdawn commented:
Yes, if I transpose `x` and `y` by `data = data.rename({'x':'y', 'y':'x'})`, it works well.
